tests/utils.py

```diff
@@ -649,7 +649,7 @@ def run(self, result=None):
         for test_name in dir(self):
             if test_name.startswith('test_'):
                 expect_failure = hasattr(self, 'not_implemented') and test_name in self.not_implemented
-                getattr(self, test_name).__dict__['__unittest_expecting_failure__'] = expect_failure
+                getattr(self, test_name).__dict__.setdefault('__unittest_expecting_failure__', expect_failure)
```

Comment:
Actually, after some further research, the value of `getattr(self, test_name).__dict__` is shared across all classes. The proper way to mark the test as 'expecting failure' in this case is by marking `__unittest_expecting_failure__` on the instance of the `TestCase`, based on the `TestCase()._testMethodName`.

I'm running a comparison of results on master and the branch to make sure they produce the same results before pushing again.